### PR TITLE
refactor: use python poetry for project management

### PR DESCRIPTION
**Pull request by cwegener**
High-level overview of changes:

1. rename src to flaresolverr in order to namespace the package
   appropriately

2. Move package.json from root folder into flaresolverr so that it can
   be read by the utils.get_flaresolverr_version() method

3. Rewrite necessary import statements to use the full package name as
   per step 1

4. Move script entry point into separate method (cli_run)

5. Manually added poetry dependencies as per existing requirements.txt
   file

### CODE REVIEW

**Comment by HLFH** [comment]
Cool thanks.

For 5, why not using [Hatch](https://github.com/pypa/hatch) instead of Poetry?

**Comment by cwegener** [comment]
@HLFH I pushed the change from poetry to hatch.

Also, I just noticed that there is a `Dockerfile` in the project ... which would also need to be updated with the changes from the renaming of `src` to `flaresolverr`

**Comment by HLFH** [comment]
I updated the `pyproject.toml` after your PR here: https://github.com/HLFH/FlareSolverr/.
I have set Python >= 3.10 because of the [updated TLS settings](https://bugs.python.org/issue43998) from the 3.10 version, and because the Dockerfile is using version 3.11.
The dynamic versioning needs to be adjusted to work.
The `package.json` file should be removed in some code around versioning should be adjusted/removed. 
I am looking into it.
I still need to update the *.txt files which I will do with `pip-compile`.

**Comment by ngosang** [comment]
@cwegener @HLFH @M4RC0Sx I don't know much about Python packaging so I have some questions.
* why hatch and no other package manager? there is a pr with pienv => #567
* what do you think about the custom logger? => #567
* what are the instructions to install the dependencies? we have to update the readme and the dockerfile

**Comment by HLFH** [comment]
https://github.com/HLFH/FlareSolverr/commits/master

I have turned off dynamic versioning until it is fixed.
requirements*.txt files are now autogenerated with `pip-compile` 

@ngosang Let me look at your questions.

**Comment by cwegener** [comment]
> @cwegener @HLFH @M4RC0Sx I don't know much about Python packaging so I have some questions.
> 
>     * why hatch and no other package manager? there is a pr with pienv => [V3beta - Python improvements #567](https://github.com/FlareSolverr/FlareSolverr/pull/567)

Here are my $.02

`hatch` is provided by the Python Packaging Authority  (PyPA), which is a core working group that provides amongst other things:
- the `pip` software
- the `pipx` software
- the Python Packaging Guide

In terms of project management tools, there is no more authoritative and mature tool than `hatch`.
https://packaging.python.org/en/latest/key_projects/#pypa-projects

> 
>     * what do you think about the custom logger? => [V3beta - Python improvements #567](https://github.com/FlareSolverr/FlareSolverr/pull/567)

Good idea. Not in scope for this change set though.

 
>     * what are the instructions to install the dependencies? we have to update the readme and the dockerfile

Dependencies are declared in the core metadata as specified in the PyPA Packaging Guide https://packaging.python.org/en/latest/specifications/core-metadata/#requires-dist-multiple-use

Any Python Source Distribution that follows the PyPA Packaging Guide should have these `Requires-Dist` fields declared in its metadata. Creating the correct metadata is handled by the project management tool, based on the configured input of installation time dependencies.

I have simply copied the contents of `requirements.txt` by hand into the respective `pyproject.toml`

As an enhancement to that, @HLFH has added the use of the `pip-compile` tool in his fork to enhance the management of those dependency declarations. `pip-compile` takes what is defined in `pyproject.toml` and builds an explicit tree of all dependencies including all transitive dependencies and stores them in one or more `requirements.txt` files.

The advantage of the `requirements.txt` file that is generated by `pip-compile` is that it has all transitive dependencies and their pinned versions, so therefore the `requirements.txt` files can then be used to perform [reproducible builds](https://reproducible-builds.org/), which may or may not be of importance depending on people's requirements.

The `requirements.txt` files are not used when performing a `pip install` of a source distribution though.

So, long story short:

For end-users, the FlareSolverr python releases should be installed from an index (PyPI is the authoritative index), using `pip install flaresolverr`

For Docker, I have no idea.

@ngosang Can you clarify what the original intention of the `Dockerfile` is? Is it so that potential contributors who want to use docker for local development can quickly launch a container? Or is the intention to use the `Dockerfile` as the basis for publishing a `flaresolverr` image to a Container registry like the Docker Hub or Quay?

**Comment by ngosang** [comment]
Thank you for the explanation. I will add all the changes from both PRs in the next release. Now I'm dealing with bug reports.

> Can you clarify what the original intention of the Dockerfile is?

This porject is distributed as Docker image with Chrome and everithing configured inside. I have 12MM downloads as today => https://hub.docker.com/r/flaresolverr/flaresolverr/
There are few users installing from source code. Mostly Windows and MacOS users until I provide compilled binaries => #660
